src/ts/ConvivaAnalytics.ts

```diff
@@ -186,89 +218,73 @@ export class ConvivaAnalytics {
-      // event. By ignoring these ad-induced events, we end up with a clean ON_SOURCE_LOADED/ON_SOURCE_UNLOADED
-      // sequence which only concerns the actual source.
-      return;
-    }
+    // Do not start a new session when a session is already existing
+    // Happens after ad playback, when the actual source is restored and an ON_SOURCE_LOADED event issued. Because
+    // we suppress the ON_SOURCE_UNLOADED event which unloads the temporary ad source, we must also ignore this
+    // event. By ignoring these ad-induced events, we end up with a clean ON_SOURCE_LOADED/ON_SOURCE_UNLOADED
+    // sequence which only concerns the actual source.
+    this.updateSession();
   };
 
   private onReady = (event: any) => {
     this.debugLog('ready', event);
-
-    let config = this.player.getConfig();
-    let autoplayEnabled = config && config.playback && config.playback.autoplay;
-
-    // Start session immediately when autoplay is enabled
-    if (autoplayEnabled) {
-      // Trigger onPlay to create a session similarly to when a user starts playback
-      this.onPlay(event);
-    }
   };
 
   private onPlaybackStateChanged = (event?: any) => {
     this.debugLog('reportplaybackstate', event);
-    let playerState = Conviva.PlayerStateManager.PlayerState.UNKNOWN;
+    let playerState;
 
-    if ((!this.player.isPlaying() && !this.player.isPaused()) || this.player.hasEnded()) {
-      // Before playback has started, and after it is finished, we report the stopped state
-      playerState = Conviva.PlayerStateManager.PlayerState.STOPPED;
-    } else if (this.player.isStalled()) {
+    if (this.player.isStalled()) {
       playerState = Conviva.PlayerStateManager.PlayerState.BUFFERING;
-    } else if (this.player.isPlaying()) {
-      playerState = Conviva.PlayerStateManager.PlayerState.PLAYING;
     } else if (this.player.isPaused()) {
       playerState = Conviva.PlayerStateManager.PlayerState.PAUSED;
+    } else if (this.player.isPlaying()) {
+      playerState = Conviva.PlayerStateManager.PlayerState.PLAYING;
     }
 
-    this.playerStateManager.setPlayerState(playerState);
+    if (playerState) {
+      this.playerStateManager.setPlayerState(playerState);
+    }
   };
 
   private onPlay = (event: any) => {
     this.debugLog('play', event);
-    if (!this.isValidSession()) {
-      if (this.isAd) {
-        this.debugLog('cannot create session during ad playback... video metadata not available');
-        return;
-      }
-
-      // Start a new session (also updates the playback state)
-      this.startSession(event);
-      // On calling play, playback is not immediately started, but the loading phase begins
-      this.playbackStarted = false;
-    } else {
-      // A normal play event happened, just update the playback state
-      this.onPlaybackStateChanged(event);
+    this.updateSession();
+    if (!this.hasPlayingEvent) {
+      this.onPlaybackStateChanged();
```

Comment:
Why `onPlaybackStateChanged()`, not `onPlaying()`?

Reply:
Because it would reach the playing state too early. When we don't have an ON_PLAYING event, we want to trigger the playing state in the first ON_TIME_CHANGED.